### ISSUE TITLE
False positive for `avoid_redundant_argument_values`

### ISSUE THREAD

**Issue by kuhnroyal** [dart-lang/linter#3103]
**Describe the issue**
After the update to Dart 2.15, I have some false warnings for generated copy code.
I boiled it down to a small sample.

In the sample below in `foo1.copyWith(a: null)` the `null` is marked as redundant because the parameter type is `String? a`.
But actual overridden implementation has a default value `String? a = _absent`.

Not sure if this is not found because of the factory delegation.
 
**To Reproduce**
```dart
void main() {
  final foo1 = Foo(a: 'test',b: 2);
  final foo2 = foo1.copyWith(a: null);


  print(foo1);
  print(foo2);
}

abstract class Foo {
  factory Foo({
    required String? a,
    required int? b,
  }) = FooImpl;

  Foo copyWith({String? a, int? b});
}


class FooImpl implements Foo {
  static const Object _absent = Object();

  final String? a;

  final int? b;

  FooImpl({
    required this.a,
    required this.b,
  });

  @override
  Foo copyWith({Object? a = _absent, Object? b= _absent,}) {
    return FooImpl(
      a: a == _absent ? this.a : a as String?,
      b: b == _absent ? this.b : b as int?,
    );
  }
}
```

**Expected behavior**
I don't expect a warning in this case.

**Comment by jamesderlin** [comment]
This is a duplicate of https://github.com/dart-lang/linter/issues/2217.

**Comment by srawlins** [comment]
As far as I can tell, this is different from #2217 as that issue was solely concerned with redirecting factory constructors. This has to do with method overrides. A minimal repro is:

```dart
void f(A a) {
  a.m(p: null);
}

class A {
  void m({String? p = null}) {}
}

class B implements A {
  void m({String? p = 'Yay'}) {}
}
```

I think this rule is fundamentally incompatible with inheritance, and the ability to change a default value. ☹️ 

@pq @bwilkerson what do you think of changing this rule to never report on arguments to instance methods? The alternative is accepting, I think, a high level of false positives.

**Comment by bwilkerson** [comment]
I suspect that the majority of invocations in most Dart code is going to be the invocation of instance methods. We could measure that to know for sure; maybe with build methods that isn't true.

I suspect that the number of times that the default value of a parameter changes in an override of the method is fairly small compared to the number of times where it doesn't. Again, we could measure to be sure.

If those are both true I would question the wisdom of disabling the lint in the majority of situations for an uncommon case and I have to question the characterization of this as a "high level of false positives". But like I said, I might be wrong about either or both of those hypotheses, so maybe the characterization is accurate.

A potential alternative would be to introduce a way to indicate which parameters are allowed to have their default value changed and to not flag in those cases (under the assumption that we don't know the default value in the method that will actually be invoked), but I haven't come up with a mechanism for marking them that doesn't introduce it's own set of problems.

**Comment by srawlins** [comment]
Yeah I think the most pragmatic solution is to close this and recommend `// ignore:`. I don't think we need to gather data on this.

https://workchronicles.substack.com/p/comic-data-driven-vs-data-informed

**Comment by srawlins** [comment]
Actually, quick note: before closing we should add a note to the lint rule's description.
